### PR TITLE
substrate-client: tests and improvements

### PR DESCRIPTION
**Pull request by josepot**
I couldn't sleep well at night knowing that we lack of comprehensive tests for the `substrate-client`, so I decided to  to begin developing them.

In the process, issues emerged in the public API, predominantly concerning error propagation, along with a few bugs.

While this PR remains a work-in-progress, it's worth noting that:

1. Tests for the `chainHead` follow operations and calls are still pending.
2. The tests solely evaluate the public API and aim to address all potential edge cases. Once complete, it will be considerably simpler to modify the intricate `substrate-client` internals.

I'm of the opinion that the underlying mechanics of this package will inevitably remain complex. Our goal should be to manage this complexity within the package, ensuring a seamless experience for users via the public API. Consequently, I'm eager to bolster the testing suite swiftly. This will not only solidify a robust public API but also enable confident refactoring to enhance maintainability.

### CODE REVIEW

**Comment by josepot** [comment]
As mentioned previously, the test development led me to identify numerous bugs and potential API enhancements. This reaffirms the importance of starting tests early in the process.

Currently, two primary areas require further tests, which I'll address in subsequent PRs:
1. **`chainHead.storage`**: I began crafting tests for this, but there's ambiguity about the expected outcomes in certain scenarios. I may need to delve deeper, perhaps even seeking clarifications from `paritytech/json-rpc-interface-spec`.
2. **Connectivity scenarios**: This includes challenges like delayed provider connectivity, and how we manage (batch) queries during these downtimes.

The testing approach revolves around creating provider fixtures and examining interactions between the public API and provider messages. I'm confident we're not encroaching on implementation specifics, ensuring these tests genuinely bolster our code simplification and maintainability objectives.

Also, the current test coverage is commendable:

```bash
 % Coverage report from v8
---------------------------|---------|----------|---------|---------|
File                       | % Stmts | % Branch | % Funcs | % Lines |
---------------------------|---------|----------|---------|---------|
All files                  |   91.95 |    90.97 |   95.08 |   91.95 |
 src                       |     100 |      100 |     100 |     100 |
  index.ts                 |     100 |      100 |     100 |     100 |
 src/chainhead             |   88.27 |    89.87 |     100 |   88.27 |
  body.ts                  |     100 |      100 |     100 |     100 |
  call.ts                  |     100 |      100 |     100 |     100 |
  chainhead.ts             |     100 |      100 |     100 |     100 |
  errors.ts                |     100 |      100 |     100 |     100 |
  header.ts                |     100 |      100 |     100 |     100 |
  index.ts                 |     100 |      100 |     100 |     100 |
  operation-promise.ts     |     100 |    93.75 |     100 |     100 |
  storage.ts               |   59.23 |    41.66 |     100 |   59.23 |
  unpin.ts                 |     100 |      100 |     100 |     100 |
 src/client                |   91.37 |    82.75 |      90 |   91.37 |
  ErrorRpc.ts              |     100 |      100 |     100 |     100 |
  createClient.ts          |   90.44 |    81.48 |    87.5 |   90.44 |
  index.ts                 |     100 |      100 |     100 |     100 |
 src/internal-utils        |     100 |      100 |   85.71 |     100 |
  abortablePromiseFn.ts    |     100 |      100 |     100 |     100 |
  deferred-promise.ts      |     100 |      100 |   33.33 |     100 |
  index.ts                 |     100 |      100 |     100 |     100 |
  noop.ts                  |     100 |      100 |     100 |     100 |
  subscriptions-manager.ts |     100 |      100 |     100 |     100 |
 src/transaction           |     100 |      100 |     100 |     100 |
  index.ts                 |     100 |      100 |     100 |     100 |
  transaction.ts           |     100 |      100 |     100 |     100 |
---------------------------|---------|----------|---------|---------|
```

File: packages/substrate-client/src/chainhead/public-types.ts [hunk 1]
```diff
@@ -86,7 +82,7 @@ export interface FollowResponse {
     StorageResponse
   >
   header: (hash: string) => Promise<string>
-  unpin: (...hashes: Array<string>) => Promise<void>
+  unpin: (hashes: Array<string>) => Promise<void>
```
Comment (josepot) [inline packages/substrate-client/src/chainhead/public-types.ts:85]: This is probably the most significant breaking change. While testing, I realized that 99.99% of the times we will be calling this method with an Array of hashes anyways... So, the spread looked fancy, but in reality is quite impractical.

File: packages/substrate-client/tests/fixtures.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+  let latestIdx = 0
+  const getNewMessages = () => {
+    const result = receivedMessages.slice(latestIdx).map((m) => JSON.parse(m))
+    latestIdx = receivedMessages.length
+    return result
+  }
+
+  const getAllMessages = () =>
+    receivedMessages.slice(0).map((m) => JSON.parse(m))
```
Comment (ryanleecode) [inline packages/substrate-client/tests/fixtures.ts:36]:
```suggestion
 const getAllMessages = () => receivedMessages.map((m) => JSON.parse(m))
```

File: packages/substrate-client/src/chainhead/chainhead.ts [hunk 3]
```diff
@@ -1,109 +1,158 @@
...
-            request("chainHead_unstable_unfollow", [subscriptionId])
-          }
 
-          followSubscription = subscriptionId
-          res(subscriptionId)
-        },
-      )
-    })
+    const deferredFollow = deferred<string | Error>()
+    let followSubscription: Promise<string | Error> | string | null =
```
Comment (ryanleecode) [inline packages/substrate-client/src/chainhead/chainhead.ts:50]: i believe we discussed this offline but this is just a hack to do streams correct.

since it starts as an unresolved promise. the first time we resolve the promise it is the "first value" in the stream. But since we cannot resolve a promise twice, we simply reassign the variable for subsequent values.

File: packages/substrate-client/tests/chainHead-functions.spec.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+      error: {
+        code: -32700,
+        message: "Parse error",
+      },
+    })
+
+    await expect(chainHead[name](...(args as [any]))).rejects.toEqual(
+      new ErrorDisjoint(),
+    )
+  })
```
Comment (ryanleecode) [inline packages/substrate-client/tests/chainHead-functions.spec.ts:105]: duplicated or intentional?

File: packages/substrate-client/tests/chainHead-functions.spec.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
+import { expect, describe, test } from "vitest"
```
Comment (ryanleecode) [inline packages/substrate-client/tests/chainHead-functions.spec.ts:1]: nit: use `it` import for all tests starting with `("it ....)`

**Review by ryanleecode — COMMENTED** [review]
A few nits but overall seems to be working nicely. Well done! :rocket:

File: packages/substrate-client/src/chainhead/errors.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
+export class ErrorStop extends Error {
```
[the remark this answers is not in the document]
Comment (josepot) [reply packages/substrate-client/src/chainhead/errors.ts:1]: Yeah... This is a good point! Thanks for pointing this out, this thing started on the wrong foot with the `ErrorRpc` and then I tried to stay consistent with that bad naming... 😬 I will change them all as you suggested.

File: packages/substrate-client/tests/fixtures.ts (hunk 2 above)
Comment (josepot) [reply packages/substrate-client/tests/fixtures.ts:36]: good catch! Yep, the `slice(0)` was a leftover b/c in the initial iteration I wasn't applying the `JSON.parse`

File: packages/substrate-client/src/transaction/transaction.ts [hunk 7]
```diff
@@ -1,29 +1,66 @@
...
+  Transaction,
+  TxFinalized,
+  TxInvalid,
+  TxDropped,
+  TxError,
+} from "./types"
 
-const finalEvents = new Set(["dropped", "invalid", "finalized"])
+type TerminalEvent = TxDropped | TxInvalid | TxFinalized | TxError
+const terminalEvents = new Set(["dropped", "invalid", "finalized", "error"])
```
[the remark this answers is not in the document]
Comment (josepot) [reply packages/substrate-client/src/transaction/transaction.ts:13]: nice!

File: packages/substrate-client/tests/chainHead-functions.spec.ts (hunk 4 above)
Comment (josepot) [reply packages/substrate-client/tests/chainHead-functions.spec.ts:105]: mmmm, if we are talking about the same thing, then it's a bit of both... I mean, specially with tests I rather having duplication than creating premature abstractions... However, I can see how it would make a lot of sense to have a `parseError` export from the fixtures, so I exported this one:

```ts
export const parseError: IRpcError = {
  code: -32700,
  message: "Parse error",
}
```

Comment (ryanleecode) [reply packages/substrate-client/tests/chainHead-functions.spec.ts:105]: lines 79-91 and 92-104 do the same thing, so the confusion is why are we doing it twice

File: packages/substrate-client/src/internal-utils/deferred-promise.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
+export interface DeferredPromise<T> {
```
[the remark this answers is not in the document]
Comment (josepot) [reply packages/substrate-client/src/internal-utils/deferred-promise.ts:1]: yeah... sorry, but I don't quite see it... The current one completely fulfils our needs, so let's stick with it until we need something fancier.

**Review by ryanleecode — APPROVED** [review]

File: packages/substrate-client/src/chainhead/chainhead.ts (hunk 3 above)
Comment (josepot) [reply packages/substrate-client/src/chainhead/chainhead.ts:50]: You're right in noting that this implementation is trying to approximate behavior typical of Observables.

The crux is not so much about being "unable to resolve a promise twice." We could retain a resolved Promise and keep leveraging it until the user opts to "unfollow" or an error occurs, subsequently setting its value to `null` to signify a dead subscription. 

However, following such an approach means any `chainHead` calls would **always** execute after a micro-task, rather than being evaluated synchronously whenever possible. Even though it sounds minor, such small deviations often lead to race conditions or unexpected consumer behaviors. Essentially, this would be a leaky abstraction. It inadvertently exposes users to the nitty-gritty of the implementation, complicating the library's usability.

Consider this scenario: If a user wishes to cancel an ongoing operation, the [`chainHead_unstable_stopOperation`](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_unstable_stopOperation.html#chainhead_unstable_stopoperation) should be invoked synchronously within that callback, not postponed to a later microtask. Always keep in mind, awaiting a resolved promise still results in a microtask, and it's never synchronous. This logic applies across all `chainHead` calls (like `unpin`, `header`, `call`, etc.). Our ultimate goal is to ensure synchronous logic execution when feasible.

Also, the variable `followSubscription` serves as a state indicator:
- As a `Promise`: Awaiting the followSubscription token; everything is pending.
- As a `string`: The `follow` subscription is in action.
- As `null`: The subscription has ended due to user cancellation, a drop event, or another error.

File: packages/substrate-client/tests/chainHead-functions.spec.ts (hunk 4 above)
Comment (josepot) [reply packages/substrate-client/tests/chainHead-functions.spec.ts]: oh, I see! yeah, I actually did that on purpose. I will add a comment. This is kinda related to [this](https://github.com/paritytech/capi/pull/65#discussion_r1309631442).

If you pay close attention, you will notice that they don't do exactly the same thing:

- lines 79-91 checks that the `DisjointError` takes place when the errored `followSubscription` message arrives **after** the user has already called `unpin` or `header`.
- lines 92-104 checks that the `DisjointError` takes place when the errored `followSubscription` message arrives **before** the user has called `unpin` or `header`.

Comment (josepot) [reply packages/substrate-client/tests/chainHead-functions.spec.ts]: I [added a comment and I also tried to make it a bit more explicit](https://github.com/paritytech/capi/pull/65/commits/a7e271e226932bb8a991a70008fab0cd9c30e404). Thanks for pointing this out!

File: packages/substrate-client/src/chainhead/chainhead.ts (hunk 3 above)
Comment (ryanleecode) [reply packages/substrate-client/src/chainhead/chainhead.ts:50]: Yup that makes sense. Lets take what you wrote here and put it as part of a readme or wiki.

File: packages/substrate-client/tests/chainHead-functions.spec.ts (hunk 4 above)
Comment (ryanleecode) [reply packages/substrate-client/tests/chainHead-functions.spec.ts]: ah yes you're right. it looked the same but it wasn't because promises are eager rather than lazy.

File: packages/substrate-client/src/chainhead/chainhead.ts (hunk 3 above)
Comment (josepot) [reply packages/substrate-client/src/chainhead/chainhead.ts:50]: Absolutely, I'm on board with integrating these insights into the library's documentation. It's crucial for users to understand not just the "how" but the "why" behind certain behaviors.

### Documentation vs. Implementation:
While documentation will provide clarity on expected behavior and its rationale, it's important that it doesn't delve into the specificities of the implementation (like I did in the previous comment). This allows for flexibility in evolving the codebase without frequent documentation overhauls.

### Tests as Documentation:
In the absence of detailed documentation, tests can serve as a practical form of documentation. I personally approach library docs with caution, often leaning on the library's tests to decipher the actual behavior. For a low-level library like this, the documentation should ideally explain the reasoning behind the test behaviors. Meaning that I would rather ironing out the tests before we get started with the docs.

### A Practical Insight:
Taking the `followSubscription` implementation as an example: someone might wonder about the re-assignment of its resolved value. If we were to comment out [this specific line](https://github.com/paritytech/capi/blob/c6c697ba18e18138f25533d443e82d78d642e5be/packages/substrate-client/src/chainhead/chainhead.ts#L85), it might seem superficially like nothing's amiss. However, tests would fail, illustrating that the public API's behavior has changed. To pass the tests again, we'd need to artificially introduce delays with constructs like `await Promise.resolve()`. This example underscores the importance of tests in elucidating library behavior.

### Moving Forward:
Given the implicit way tests can document behaviors, it's all the more essential to ensure our documentation explicitly describes both what the library does and why it does it. Ensuring transparency and clarity in our design rationale is key.

So, yes 💯 : this behavior (and the reasoning behind it) should be explicitly documented, but "one step at a time." 😊
